pkg/controller/broker_client_manager.go

```diff
@@ -0,0 +1,132 @@
+func (bk *BrokerKey) IsClusterScoped() bool {
+	return bk.namespace == ""
+}
+
+// String returns string representation of the broker key
+func (bk *BrokerKey) String() string {
+	if bk.IsClusterScoped() {
+		return bk.name
+	}
+	return fmt.Sprintf("%s/%s", bk.namespace, bk.name)
+}
+
+// NewServiceBrokerKey creates a BrokerKey instance which points to namespaced broker
+func NewServiceBrokerKey(namespace, name string) BrokerKey {
+	return BrokerKey{
+		namespace: namespace,
+		name:      name,
+	}
+}
+
+// NewClusterServiceBrokerKey creates a BrokerKey instance which points to cluster wide broker
+func NewClusterServiceBrokerKey(name string) BrokerKey {
+	return BrokerKey{
+		namespace: "",
+		name:      name,
+	}
+}
+
+// BrokerClientManager stores OSB client instances per broker
+type BrokerClientManager struct {
+	mu      sync.RWMutex
+	clients map[BrokerKey]clientWithConfig
+
+	brokerClientCreateFunc osb.CreateFunc
+}
+
+// NewBrokerClientManager creates BrokerClientManager instance
+func NewBrokerClientManager(brokerClientCreateFunc osb.CreateFunc) *BrokerClientManager {
+	return &BrokerClientManager{
+		clients:                map[BrokerKey]clientWithConfig{},
+		brokerClientCreateFunc: brokerClientCreateFunc,
+	}
+}
+
+// UpdateBrokerClient creates new broker client if necessary (the ClientConfig has changed or there is no client for the broker),
+// the method returns created or stored osb.Client instance.
+func (m *BrokerClientManager) UpdateBrokerClient(brokerKey BrokerKey, clientConfig *osb.ClientConfiguration) (osb.Client, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	existing, found := m.clients[brokerKey]
+
+	if !found || m.configHasChanged(existing.clientConfig, clientConfig) {
+		glog.V(4).Infof("Updating OSB client for broker %s, URL: %s", brokerKey.String(), clientConfig.URL)
+		return m.createClient(brokerKey, clientConfig)
+	}
+
+	return existing.OSBClient, nil
+}
+
+// RemoveBrokerClient removes broker client broker
+func (m *BrokerClientManager) RemoveBrokerClient(brokerKey BrokerKey) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	glog.V(4).Info("Removing OSB client for broker %s", brokerKey.String())
+	delete(m.clients, brokerKey)
+}
+
+// BrokerClient returns broker client fro a broker specified by the brokerKey
+func (m *BrokerClientManager) BrokerClient(brokerKey BrokerKey) (osb.Client, bool) {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	existing, found := m.clients[brokerKey]
+	return existing.OSBClient, found
+}
+
+func (m *BrokerClientManager) createClient(brokerKey BrokerKey, clientConfig *osb.ClientConfiguration) (osb.Client, error) {
+	client, err := m.brokerClientCreateFunc(clientConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	m.clients[brokerKey] = clientWithConfig{
+		OSBClient:    client,
+		clientConfig: clientConfig,
+	}
+	return client, nil
+}
+
+func (m *BrokerClientManager) configHasChanged(cfg1 *osb.ClientConfiguration, cfg2 *osb.ClientConfiguration) bool {
```

Comment:
Would this be better if it was a function instead of a method?

Reply:
I wrote the helper method to be used inside BrokerClientManager and I don't want to expose it. It was not designed to be used by other components.